c-dependencies/js-compute-runtime/js-compute-builtins.cpp

```diff
@@ -694,11 +694,287 @@ bool parse_body(JSContext *cx, HandleObject self, UniqueChars buf, size_t len) {
+
+  auto reader_val = js::GetFunctionNativeReserved(catch_handler, 1);
+  MOZ_ASSERT(reader_val.isObject());
+  RootedObject reader(cx, &reader_val.toObject());
+
+  // We're guaranteed to work with a native ReadableStreamDefaultReader here as we used
+  // `JS::ReadableStreamDefaultReaderRead(cx, reader)`, which in turn is guaranteed to return {done:
+  // bool, value: any} objects to read promise then callbacks.
+  MOZ_ASSERT(args[0].isObject());
+  RootedObject chunk_obj(cx, &args[0].toObject());
+  RootedValue done_val(cx);
+  RootedValue value(cx);
+#ifdef DEBUG
+  bool hasValue;
+  if (!JS_HasProperty(cx, chunk_obj, "value", &hasValue)) {
+    return false;
+  }
+  MOZ_ASSERT(hasValue);
+#endif
+  if (!JS_GetProperty(cx, chunk_obj, "value", &value)) {
+    return false;
+  }
+#ifdef DEBUG
+  bool hasDone;
+  if (!JS_HasProperty(cx, chunk_obj, "done", &hasDone)) {
+    return false;
+  }
+  MOZ_ASSERT(hasDone);
+#endif
+  if (!JS_GetProperty(cx, chunk_obj, "done", &done_val)) {
+    return false;
+  }
+  MOZ_ASSERT(done_val.isBoolean());
+  if (done_val.toBoolean()) {
+    // We finished reading the stream
+    // Now we need to iterate/reduce `contents` JS Array into UniqueChars
+    uint32_t contentsLength;
+    if (!JS::GetArrayLength(cx, contents, &contentsLength)) {
+      return false;
+    }
+    // TODO(performance): investigate whether we can infer the size directly from `contents`
+    size_t buf_size = HANDLE_READ_CHUNK_SIZE;
+    // TODO(performance): make use of malloc slack.
+    // https://github.com/fastly/js-compute-runtime/issues/217
+    char *buf = static_cast<char *>(JS_malloc(cx, buf_size));
+    if (!buf) {
+      JS_ReportOutOfMemory(cx);
+      return false;
+    }
+    // For realloc below.
+    char *new_buf;
+    size_t offset = 0;
+    for (uint32_t index = 0; index < contentsLength; index++) {
+      RootedValue val(cx);
+      if (!JS_GetElement(cx, contents, index, &val)) {
+        JS_free(cx, buf);
+        return false;
+      }
+      {
+        JS::AutoCheckCannotGC nogc;
+        MOZ_ASSERT(val.isObject());
+        JSObject *array = &val.toObject();
+        MOZ_ASSERT(JS_IsUint8Array(array));
+        bool is_shared;
+        static_assert(CHAR_BIT == 8, "Strange char");
+        char *bytes = reinterpret_cast<char *>(JS_GetUint8ArrayData(array, &is_shared, nogc));
+        size_t length = JS_GetTypedArrayByteLength(array);
+        offset += length;
+        if (length + offset > buf_size) {
+          while (length + offset > buf_size) {
+            size_t new_size = buf_size + HANDLE_READ_CHUNK_SIZE;
+            new_buf = static_cast<char *>(JS_realloc(cx, buf, buf_size, new_size));
+            if (!new_buf) {
+              JS_free(cx, buf);
+              JS_ReportOutOfMemory(cx);
+              return false;
+            }
+            buf = new_buf;
+            buf_size += HANDLE_READ_CHUNK_SIZE;
+          }
+        }
+        // strcat(buf, *bytes);
+        memcpy(buf, bytes, length);
```

Comment:
This was necessary as it was the only place in the function where we were putting the bytes into buf.

memcpy was very incorrect to use though as it would replace the buf contents instead of concatenating into buf.

I've updated to use strcpy and strcat instead and will add our own custom tests to ensure this functionality is correctly implemented as it seems web-platform-tests does not cover this